java/arcs/core/util/Scheduler.kt

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.util
+
+import kotlin.coroutines.CoroutineContext
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.update
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.TimeoutCancellationException
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.withTimeout
+
+/**
+ * The [Scheduler] is responsible for scheduling the execution of a batch of [Task]s (known as an
+ * [Agenda]), at a defined maximum [scheduleRateHz], where each agenda is limited to running within
+ * the specified [agendaProcessingTimeout] window.
+ */
+class Scheduler(
+    private val time: Time,
+    context: CoroutineContext,
+    private val agendaProcessingTimeout: Long = DEFAULT_AGENDA_PROCESSING_TIMEOUT,
+    private val scheduleRateHz: Int = DEFAULT_SCHEDULE_RATE_HZ
+) {
+    private val log = TaggedLog { "Scheduler(${hashCode()})" }
+    /* internal */
+    val launches = atomic(0)
+    /* internal */
+    val loops = atomic(0)
+    private var processingJob: Job? = null
+    private val scope = CoroutineScope(context)
+    private val isIdle = atomic(true)
+    private val isPaused = atomic(false)
+    private val agenda = atomic(Agenda())
+
+    /** Schedule a single [Task] to be run as part of the next agenda. */
+    fun schedule(task: Task) {
+        agenda.update { it.addTask(task) }
+        if (isIdle.value) scope.startProcessingJob()
+    }
+
+    /** Schedule some [Task]s to be run as part of the next agenda. */
+    fun schedule(tasks: Iterable<Task>) {
+        agenda.update { it.addTasks(tasks) }
+        if (isIdle.value) scope.startProcessingJob()
+    }
+
+    /** Pause evaluation of the agenda. */
+    fun pause() {
+        isPaused.value = true
+    }
+
+    /** Resume evaluation of scheduled [Task]s in the agenda. */
+    fun resume() {
+        if (!isPaused.compareAndSet(expect = true, update = false)) return
+        // If we were paused, we can re-start.
+        scope.startProcessingJob()
+    }
+
+    /** Wait for the [Scheduler] to become idle. */
+    suspend fun waitForIdle() {
+        processingJob?.join()
+    }
+
+    private fun CoroutineScope.startProcessingJob() {
+        isIdle.value = false
+        processingJob = launch {
+            launches.incrementAndGet()
+
+            var shallContinue = true
+            while (shallContinue) {
+                val startTime = time.currentTimeMillis
+                shallContinue = withTimeout(agendaProcessingTimeout) {
+                    suspendCancellableCoroutine<Boolean> {
```

Comment:
What is wrapping this in `suspendCancellableCoroutine` providing us here?

Reply:
This allows the `withTimeout` to tell the `process` block it's been canceled. I think it'll be more powerful if I push the suspendCancelableCoroutine/continuation stuff deeper, so we can check for cancellation between each processor/listener.

Reply:
Pushed it a bit deeper, and am now checking whether or not the job has been canceled between each scheduled task.

Reply:
Nice addition!